build-system/babel-plugins/babel-plugin-jsx-style-object/index.js

```diff
@@ -0,0 +1,181 @@
+/**
+ * @fileoverview
+ * Converts Object Expression syntax in `style={{foo: 'bar'}}` to a concatenated
+ * string expression.
+ *
+ * This only transforms a file if it imports the `baseModule` for JSX below,
+ * which lacks runtime support for converting the Object into a string.
+ * This transform takes that responsibility instead.
+ */
+
+const baseModule = 'core/dom/jsx';
+const helperModule = 'core/dom/jsx-style-property-string';
+const helperFn = 'jsxStylePropertyString';
+
+const nonDimensional = new Set([
+  'animation-iteration-count',
+  'box-flex',
+  'box-flex-group',
+  'box-ordinal-group',
+  'column-count',
+  'fill-opacity',
+  'flex',
+  'flex-grow',
+  'flex-positive',
+  'flex-shrink',
+  'flex-negative',
+  'flex-order',
+  'font-weight',
+  'line-clamp',
+  'line-height',
+  'opacity',
+  'order',
+  'orphans',
+  'stop-opacity',
+  'stroke-dashoffset',
+  'stroke-opacity',
+  'stroke-width',
+  'tab-size',
+  'widows',
+  'z-index',
+  'zoom',
+]);
+
+module.exports = function (babel) {
+  const {types: t} = babel;
+
+  const dashCase = (camelCase) =>
+    camelCase.replace(/[A-Z]/g, '-$&').toLowerCase();
+
+  let hasBaseModule = false;
+  let hasHelperModule = false;
+  let programPath = null;
+
+  /**
+   * @param {babel.NodePath<import('@babel/types').ObjectProperty>} path
+   * @return {?babel.Node}
+   */
+  function transformProp(path) {
+    // @ts-ignore
+    const name = path.node.key.name || path.node.key.value;
+    const cssName = dashCase(name);
+    const isDimensional = !nonDimensional.has(cssName);
+    const value = path.get('value');
+    const evaluated = value.evaluate();
+
+    // If we can evaluate the value, return a composed string.
+    if (evaluated.confident) {
+      const {value} = evaluated;
+      if (value == null || value === '') {
+        return null;
+      }
+      const withUnit =
+        isDimensional && typeof value === 'number' ? `${value}px` : value;
+      return t.stringLiteral(`${cssName}:${withUnit};`);
+    }
+
+    // Otherwise call the helper function to evaluate nullish and dimensional values.
+    if (!hasHelperModule && programPath) {
+      hasHelperModule = true;
+      programPath.unshiftContainer(
+        'body',
+        t.importDeclaration(
+          [t.importSpecifier(t.identifier(helperFn), t.identifier(helperFn))],
+          t.stringLiteral(helperModule)
+        )
+      );
```

Comment:
Nit: prefer `@babel/helper-module-imports`'s `addNamed`, and let it run unconditionally.

Reply:
Are we okay with this?
![image](https://user-images.githubusercontent.com/254946/141229285-adf4c9d2-1568-4db5-b5bb-65f89f4c7375.png)